aws/resource_aws_synthetics_canary.go

```diff
@@ -0,0 +1,649 @@
+package aws
+
+import (
+	"fmt"
+	"log"
+	"strings"
+	"time"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/arn"
+	"github.com/aws/aws-sdk-go/service/synthetics"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/validation"
+	"github.com/terraform-providers/terraform-provider-aws/aws/internal/keyvaluetags"
+	"github.com/terraform-providers/terraform-provider-aws/aws/internal/service/synthetics/waiter"
+)
+
+const awsMutexCanary = `aws_synthetics_canary`
+
+func resourceAwsSyntheticsCanary() *schema.Resource {
+	return &schema.Resource{
+		Create: resourceAwsSyntheticsCanaryCreate,
+		Read:   resourceAwsSyntheticsCanaryRead,
+		Update: resourceAwsSyntheticsCanaryUpdate,
+		Delete: resourceAwsSyntheticsCanaryDelete,
+		Importer: &schema.ResourceImporter{
+			State: schema.ImportStatePassthrough,
+		},
+
+		Schema: map[string]*schema.Schema{
+			"name": {
+				Type:         schema.TypeString,
+				Required:     true,
+				ForceNew:     true,
+				ValidateFunc: validation.StringLenBetween(1, 21),
+			},
+			"runtime_version": {
+				Type:     schema.TypeString,
+				Required: true,
+			},
+			"artifact_s3_location": {
+				Type:     schema.TypeString,
+				Required: true,
+				ForceNew: true,
+				DiffSuppressFunc: func(k, old, new string, d *schema.ResourceData) bool {
+					return strings.TrimPrefix(new, "s3://") == old
+				},
+			},
+			"handler": {
+				Type:     schema.TypeString,
+				Required: true,
+			},
+			"s3_bucket": {
+				Type:          schema.TypeString,
+				Optional:      true,
+				ConflictsWith: []string{"zip_file"},
+				RequiredWith:  []string{"s3_key"},
+			},
+			"s3_key": {
+				Type:          schema.TypeString,
+				Optional:      true,
+				ConflictsWith: []string{"zip_file"},
+				RequiredWith:  []string{"s3_bucket"},
+			},
+			"s3_version": {
+				Type:          schema.TypeString,
+				Optional:      true,
+				ConflictsWith: []string{"zip_file"},
+			},
+			"zip_file": {
+				Type:          schema.TypeString,
+				Optional:      true,
+				ConflictsWith: []string{"s3_bucket", "s3_key", "s3_version"},
+			},
```

Comment:
Should live in a `code` configuration block to match the [AWS API](https://docs.aws.amazon.com/AmazonSynthetics/latest/APIReference/API_CreateCanary.html#API_CreateCanary_RequestSyntax).

Reply:
Cant do that, getting diffs as these arguments are not returned by API. this is handled similarly to how to lambda functions. there is a `FunctionCode` object in the API but its in the root of the resource schema.